doc: Update readme (#16)
gnoswap-labs/governance: README.md

> # governance

Reviewer: ```suggestion
# Governance
```

>:warning: This package is under active development and yet audited for security. Use at your own risk.

> The governance system is a comprehensive decentralized solution for managing proposals, voting and executing decisions in a a environment. It is designed to be gno platform.

Reviewer: ```suggestion
This repository contains realms (smart-contracts) for token-based governance on gno.land. Any gno.land-based tokens can use `Airdrop` and 'Snapshot` for distributing tokens, and use `Staker`, `Delegation, and `Governor` for governance. All realms are customizable, meaning they can be used together, or independently. 
```

> ## Key Components
> 
> 1. **Governor**: Manages the proposal's lifecycle, including creation, voting, and execution. It also handles the governance configuration and upgrade.
> 
> 2. **Staker**: Allows users to stake tokens and delegate voting power. It maintains a history of delegation data to support time-weighted voting.
> 
> 3. **Snapshot**: Provides a Cairo's snapshot functionality. It captures the state of the system at specific points in time.
> 
> 4. **Airdrop**: handles token distribution events, including the claiming process and refund if necessary.
> 
> 5. **Delegation**: Allows token holders to delegate their voting power without transferring token ownership, increasing governance participation flexibility.

Reviewer: ```suggestion
## Key Components

1. **Snapshot**: `Snapshot` provides capturing, storing, and retrieving snapshots of the state of a system at specific points in time, including claimed bitmaps and the number of remaining tokens.

2. **Airdrop**: `Airdrop` handles token distribution events, including the claiming process and refund if necessary.

3. **Staker**: `Staker` allows users to stake tokens and delegate voting power. It maintains a history of delegation data to support time-weighted voting.

4. **Delegation**: `Delegation` allows token holders to delegate their voting power without transferring token ownership, increasing governance participation flexibility.

5. **Governor**: `Governor` manages the proposal's lifecycle, including creation, voting, and execution. It also handles the governance configuration and upgrades.
```

## System Overview

The following diagram illustrates hw the main components of the governance system interact with each other:

```mermaid
graph TD
    A[User] -->|Interacts| B(Governor)
    B -->|Manages Proposals| B
    B -->|Executes Decisions| B
    B -->|Uses| C(Staker)
    B -->|Creates/Uses| D(Snapshot)
    C -->|Manages Stakes| C
    C -->|Records History| C
    E[Token Holder] -->|Stakes/Delegates| C
    F[Airdrop Participant] -->|Claims| G(Airdrop)
    G -->|Checks Eligibility| D
    H[Delegator] -->|Delegates Voting Power| I(Delegation)
    I -->|Updates| C
    B -->|Checks Voting Power| I

> ```

Reviewer: Hmm it seems quite complicated and not easy to understand at the first look.

In order for a user to interact with the system, this would be the order I think:
1. stake
2. delegate
3. governor
4. voting

Can you improve it?

Author: modified it like below, what do you think about it?

```mermaid
sequenceDiagram
    participant U as User
    participant T as Token Contract
    participant S as Staker
    participant D as Delegation
    participant G as Governor
    participant SN as Snapshot

    Note over U,SN: 1. Stake
    U->>T: Approve tokens
    U->>S: Stake tokens
    S->>T: Transfer tokens
    S->>SN: Create stake snapshot

    Note over U,SN: 2. Delegate
    U->>D: Delegate voting power
    D->>S: Update delegation info
    S->>SN: Create delegation snapshot

    Note over U,SN: 3. Governor
    U->>G: Create proposal
    G->>SN: Capture proposal state
    G->>S: Check proposer's stake

    Note over U,SN: 4. Voting
    U->>G: Cast vote
    G->>S: Check voter's stake/delegation
    G->>SN: Record vote
    G->>G: votes
    G->>G: Execute proposal (if passed)
```

Reviewer: Much much better! thanks.

> ## Features
> 
> - Proposal creation and management
> - Configurable governance parameters
> - Upgradable contract design

Reviewer: ```suggestion
## Features

- Distribution gno.land based tokens with `Snapshot` and `Airdrop`
- Creation of a governance system for gno.land based tokens with `Staker`, `Delegation`, and `Governor`
- Proposal creation and management
- Configurable governance parameters
- Providing query functions for delegators such as current delegation, historical delegation, average delegation, and cumulative delegation
- Upgradable contract design
```

# Governance

> >:warning: This package is under active development and yet audited for security. Use at your own risk.
> 
> This repository contains realms (smart-contracts) for token-based governance on gno.land. Any gno.land-based tokens can use `Airdrop` and 'Snapshot` for distributing tokens, and use `Staker`, `Delegation, and `Governor` for governance. All realms are customizable, meaning they can be used together, or independently. 

Reviewer: ```suggestion
>:warning: This package is under active development and yet audited for security. Use at your own risk.

This repository contains realms (smart-contracts) for token-based governance on gno.land. Any gno.land-based tokens can use `Airdrop` and `Snapshot` for distributing tokens, and use `Staker`, `Delegation`, and `Governor` for governance. All realms are customizable, meaning they can be used together, or independently. 
```

3. **Staker**: `Staker` allows users to stake tokens and delegate voting power. It maintains a history of delegation data to support time-weighted voting.

4. **Delegation**: `Delegation` allows token holders to delegate their voting power without transferring token ownership, increasing governance participation flexibility.

5. **Governor**: `Governor` manages the proposal's lifecycle, including creation, voting, and execution. It also handles the governance configuration and upgrades.

## System Overview

The following diagram illustrates hw the main components of the governance system interact with each other:

```mermaid
graph TD
    A[User] -->|Interacts| B(Governor)
    B -->|Manages Proposals| B
    B -->|Executes Decisions| B
    B -->|Uses| C(Staker)
    B -->|Creates/Uses| D(Snapshot)
    C -->|Manages Stakes| C
    C -->|Records History| C
    E[Token Holder] -->|Stakes/Delegates| C
    F[Airdrop Participant] -->|Claims| G(Airdrop)
    G -->|Checks Eligibility| D
    H[Delegator] -->|Delegates Voting Power| I(Delegation)
    I -->|Updates| C
    B -->|Checks Voting Power| I
```

> ## Features
> 
> - Distribution gno.land based tokens with `Snapshot` and `Airdrop`
> - Creation of a governance system for gno.land based tokens with `Staker`, `Delegation`, and `Governor`
> - Proposal creation and management
> - Configurable governance parameters
> - Providing query functions for delegators such as current delegation, historical delegation, average delegation, and cumulative delegation
> - Upgradable contract design

Reviewer: ```suggestion
## Features

- Distribution of gno.land based tokens with `Snapshot` and `Airdrop`
- Creation of a governance system for gno.land based tokens with `Staker`, `Delegation`, and `Governor`
- Proposal creation and management
- Configurable governance parameters
- Providing query functions for delegation such as current delegation, historical delegation, average delegation, and cumulative delegation
- Upgradable contract design
```
